python_modules/dagster/dagster/_core/definitions/multi_dimensional_partitions.py

```diff
@@ -246,43 +252,95 @@ def get_multi_partition_key_from_str(self, partition_key_str: str) -> MultiParti
-        for idx, key in enumerate(partition_key_strs):
-            check.invariant(
-                key in keys_per_dimension[idx][1],
-                f"Partition key {key} not found in dimension {partition_key_dims_by_idx[idx][0]}",
-            )
 
-        multi_partition_key = MultiPartitionKey(
-            {partition_key_dims_by_idx[idx]: key for idx, key in enumerate(partition_key_strs)}
+        return MultiPartitionKey(
+            {dim.name: partition_key_strs[i] for i, dim in enumerate(self._partitions_defs)}
         )
-        return multi_partition_key
 
-    def deserialize_subset(self, serialized: str) -> "MultiPartitionsSubset":
+    def deserialize_subset(self, serialized: str) -> "PartitionsSubset":
         return MultiPartitionsSubset.from_serialized(self, serialized)
 
+    def supports_deserialization(self, serialized: str) -> bool:
+        return MultiPartitionsSubset.supports_deserialization(serialized)
+
+    def _get_primary_and_secondary_dimension(
+        self,
+    ) -> Tuple[PartitionDimensionDefinition, PartitionDimensionDefinition]:
+        # Multipartitions subsets are serialized by primary dimension. If changing
+        # the selection of primary/secondary dimension, will need to also update the
+        # serialization of MultiPartitionsSubsets
+
+        time_dimensions = [
+            dim
+            for dim in self.partitions_defs
+            if isinstance(dim.partitions_def, TimeWindowPartitionsDefinition)
+        ]
+        if len(time_dimensions) == 1:
+            primary_dimension, secondary_dimension = time_dimensions[0], next(
+                iter([dim for dim in self.partitions_defs if dim != time_dimensions[0]])
+            )
+        else:
+            primary_dimension, secondary_dimension = (
+                self.partitions_defs[0],
+                self.partitions_defs[1],
+            )
+
+        return primary_dimension, secondary_dimension
+
+    @property
+    def primary_dimension(self) -> PartitionDimensionDefinition:
+        return self._get_primary_and_secondary_dimension()[0]
+
+    @property
+    def secondary_dimension(self) -> PartitionDimensionDefinition:
+        return self._get_primary_and_secondary_dimension()[1]
+
 
 class MultiPartitionsSubset(DefaultPartitionsSubset):
-    @staticmethod
-    def from_serialized(
-        partitions_def: PartitionsDefinition, serialized: str
-    ) -> "MultiPartitionsSubset":
-        if not isinstance(partitions_def, MultiPartitionsDefinition):
-            check.failed(
-                "Must pass a MultiPartitionsDefinition object to deserialize MultiPartitionsSubset."
+    def __init__(
+        self,
+        partitions_def: MultiPartitionsDefinition,
+        subset: Optional[Set[str]] = None,
+    ):
+        """
+        Users should not directly call this method. Instead, call partitions_def.empty_subset().with_partition_keys().
```

Comment:
`with_partition_keys` will validate that the partition keys exist in the partitions definition. However, when we deserialize a partitions subset and call `from_serialized`, we don't want the constructor to have to recheck that all of the partitions are valid. So I thought a good way of confirming that all partitions in the subset are valid would be to recommend that users not directly call the constructor and instead call `with_partition_keys`.

Reply:
I see. It's a little confusing here what's meant by "users", because this isn't even publicly exposed, right? And it's difficult for someone who has a `MultiPartitionsSubset` to know whether the invariant applies, because it depends on how it was constructed. IMO we should either enforce the invariant or not enforce it vs. only enforcing it in some situations.

If we do want to enforce it and we're worried about performance, one way to address that could be to cache the set of partition keys.

Reply:
I agree that it's awkward, the struggle is that if we enforced this constraint in the constructor we'd have to incur a 9s validation cost for 100K time partitions just to deserialize data we're already sure is valid.

I'm just going to remove the checks and rely on the asset status cache logic to validate that the partitions are valid.